Fix badge disappearing bug

vedanshujain opened the pull request:
We currently assume that a profile will be associated with only single WordCamp, and handle adding and removing badges (speaker/organizer) with this assumption.
However, this assumption is incorrect and causes badge disappearing bug, as described in 703.

This patch refactors the participant-notifier class such that:

1. Whenever a badge is added, an entry is made in to usermeta table, with meta key as `wc_{post_type}_{blog_id}_{post_id}`. This has following benefits:
	a. Usermeta table is shared across all WordCamp, so this can act as a central place to record whenever a badge is added for a user.
	b. Since meta key columns is indexed, a count query can be done on `wc_{post_type}_%` with `user_id = {user_id}` clause.
		Because of how MySQL stores indexes, prefix queries are also as performant as full text queries. This gives us ability to find out in how many WordCamp a user is registered as a speaker/organizer using a single performant query.
		We can use the results of this query to decide if we want to remove a badge or not.

2. In block editor, as per WordPress/gutenberg#12897, POST data will not be there when publishing a draft post. This patch adds a workaround by checking if we need to add/remove badge even if both previous and next status are `publish`.

3. Uses a meta data to prevent duplicate activity log request. This is necessary because of workaround described in above patch.

coreymckrill commented on public_html/wp-content/plugins/wordcamp-participation-notifier/wordcamp-participation-notifier.php:
```diff
@@ -156,19 +150,128 @@ protected function published_speaker_post_updated( $post ) {
+	protected function add_badge( $post, $user_id ) {
+		if ( ! in_array( $post->post_type, [ 'wcb_speaker', 'wcb_organizer' ] ) || ! $user_id ) {
+			return;
+		}
```
Could the `is_post_notifiable` method be used here instead? That would make this DRYer and easier to refactor if we ever added another post type.

vedanshujain replied:
:+1: Fixed in https://github.com/WordPress/wordcamp.org/pull/125/commits/d4d2d112f6e4d8083cefa51c327d12579e8970d7

coreymckrill commented on public_html/wp-content/plugins/wordcamp-participation-notifier/wordcamp-participation-notifier.php:
```diff
@@ -156,19 +150,128 @@ protected function published_speaker_post_updated( $post ) {
+
+		if ( ! in_array( $post->post_type, [ 'wcb_speaker', 'wcb_organizer' ] ) ) {
+			return;
+		}
```
Could the `is_post_notifiable` method be used here instead? That would make this DRYer and easier to refactor if we ever added another post type.

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/125/commits/d4d2d112f6e4d8083cefa51c327d12579e8970d7

coreymckrill commented on public_html/wp-content/plugins/wordcamp-participation-notifier/wordcamp-participation-notifier.php:
```diff
@@ -112,37 +112,31 @@ protected function is_post_notifiable( $post ) {
 	 * Updates the activity and associations of a profile when the WordPress.org username on a published speaker
 	 * or organizer post changes.
 	 *
+	 * IMPORTANT NOTE: If a draft post is published via block editor, we will have to add badges and activity here instead of `post_published_or_unpublished` method.
```
```suggestion
	 * IMPORTANT NOTE: When a draft post is published via the block editor, badges and activity must be managed here instead of in the `post_published_or_unpublished` method.
```

vedanshujain replied:
Fixed in https://github.com/WordPress/wordcamp.org/pull/125/commits/d4d2d112f6e4d8083cefa51c327d12579e8970d7

coreymckrill reviewed (commented):
I didn't test the code locally, but this looks great. A few minor questions/suggestions below.

vedanshujain commented:
- [ ]  Back fill script to add metadata for existing speaker/organizers

vedanshujain committed:
WordCamp: Rename methods to reduce ambiguity.

The method `published_speaker_post_updated` and `speaker_post_published_or_unpublished` actually effects both `speaker` and `organizer` post types, instead of just `speaker`.

This patch renames them to remove `speaker` keyword to reduce confusion.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8906 74240141-8908-4e6f-9713-ba540dce6ec7

vedanshujain committed:
WC: Refactor participant notifier to fix design flaw. See 703. See desc.

We currently assume that a profile will be associated with only single WordCamp, and handle adding and removing badges (speaker/organizer) with this assumption.
However, this assumption is incorrect and causes badge disappearing bug, as described in 703.

This patch refactors the participant-notifier class such that:

1. Whenever a badge is added, an entry is made in to usermeta table, with meta key as wc_{post_type}_{blog_id}_{post_id]. This has following benefits:
	a. Usermeta table is shared across all WordCamp, so this can act as a central place to record whenever a badge is added for a user.
	b. Since meta key columns is indexed, a count query can be done on `wc_{post_type}_%` with `user_id = {user_id}` clause.
		Because of how MySQL stores indexes, prefix queries are also as performant as full text queries. This gives us ability to find out in how many WordCamp a user is registered as a speaker/organizer using a single performant query.
		We can use the results of this query to decide if we want to remove a badge or not.
2. In block editor, as per https://github.com/WordPress/gutenberg/issues/12897, POST data will not be there when publishing a draft post. This patch adds a workaround by checking if we need to add/remove badge even if both previous and next status are `publish`.
3. Uses a meta data to prevent duplicate activity log request. This is necessary because of workaround described in above patch.

Fixes 703

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8907 74240141-8908-4e6f-9713-ba540dce6ec7

vedanshujain committed:
WC Participant Notifier: Code review fixes. See desc:

1. Use `is_post_notifiable` to make code DRY
2. Better descriptive code comment.

git-svn-id: https://meta.svn.wordpress.org/sites/trunk/wordcamp.org@8908 74240141-8908-4e6f-9713-ba540dce6ec7
